ownCloud/Settings/Passcode/PasscodeManager.swift

```diff
@@ -0,0 +1,375 @@
+//
+//  PasscodeManager.swift
+//  ownCloud
+//
+//  Created by Javier Gonzalez on 06/05/2018.
+//  Copyright © 2018 ownCloud GmbH. All rights reserved.
+//
+
+/*
+ * Copyright (C) 2018, ownCloud GmbH.
+ *
+ * This code is covered by the GNU Public License Version 3.
+ *
+ * For distribution utilizing Apple mechanisms please see https://owncloud.org/contribute/iOS-license-exception/
+ * You should have received a copy of this license along with this program. If not, see <http://www.gnu.org/licenses/gpl-3.0.en.html>.
+ *
+ */
+
+import UIKit
+import ownCloudSDK
+
+class PasscodeManager: NSObject {
+
+    // MARK: - Interface view mode
+    enum PasscodeInterfaceMode {
+        case addPasscodeFirstStep
+        case addPasscodeSecondStep
+        case unlockPasscode
+        case unlockPasscodeError
+        case deletePasscode
+        case deletePasscodeError
+        case addPasscodeFirstStepAfterErrorOnSecond
+    }
+
+    // MARK: Global vars
+
+    // Common
+    private var window: LockWindow?
+    private let passcodeLength = 4
+    private let passcodeKeychainAccount = "passcode-keychain-account"
+    private let passcodeKeychainPath = "passcode-keychain-path"
+    private var passcodeMode: PasscodeInterfaceMode?
+    private var passcodeViewController: PasscodeViewController?
+
+    // Add/Delete
+    private var passcodeFromFirstStep: String?
+    private var completionHandler: CompletionHandler?
+
+    // Unlock
+    private var datePressedHomeButton: Date?
+    private var userDefaults: UserDefaults?
+
+    // Brute force protection
+    public let TimesPasscodeFailedKey: String =  "times-passcode-failed"
+    public let DateAllowTryPasscodeAgainKey: String =  "date-allow-try-passcode-again"
+    private var timesPasscodeFailed: Int {
+        didSet {
+            self.userDefaults!.set(timesPasscodeFailed, forKey: TimesPasscodeFailedKey)
+        }
+    }
+    private var dateAllowTryAgain: Date? {
+        didSet {
+            self.userDefaults!.set(NSKeyedArchiver.archivedData(withRootObject: dateAllowTryAgain as Any), forKey: DateAllowTryPasscodeAgainKey)
+        }
+    }
+    private let timesAllowPasscodeFail: Int = 3
+    private let multiplierBruteForce: Int = 10
+    private var timerBruteForce: Timer?
+
+    // Utils
+    var isPasscodeActivated: Bool {
+        return (self.userDefaults!.bool(forKey: SecuritySettingsPasscodeKey) && isPasscodeStoredOnKeychain)
+    }
+
+    var isPasscodeStoredOnKeychain: Bool {
```

Comment:
you already have the variable `passcodeFromKeychain: String?` so this method is redundant when you need to see if the passcode is or not in the keychain you could ask this var and see its content:
- **nil**: there is no passcode stored in the keychain.
- **value**: you have the value of the passcode.

Reply:
I have both methods for two reasons:
- `isPasscodeStoredOnKeychain` is faster than `passcodeFromKeychain` because you dot not need make any conversion.
- `isPasscodeStoredOnKeychain` is public and `passcodeFromKeychain`. I do not want to make public a method to get the passcode value from outside this class.